astropy/io/fits/hdu/base.py

```diff
@@ -1464,55 +1464,23 @@ def _compute_checksum(self, data, sum32=0):
         -------
         ones complement checksum
         """
-        blocklen = 2880
-        sum32 = np.uint32(sum32)
-        for i in range(0, len(data), blocklen):
-            length = min(blocklen, len(data) - i)  # ????
-            sum32 = self._compute_hdu_checksum(data[i : i + length], sum32)
-        return sum32
-
-    def _compute_hdu_checksum(self, data, sum32=0):
-        """
-        Translated from FITS Checksum Proposal by Seaman, Pence, and Rots.
-        Use uint32 literals as a hedge against type promotion to int64.
-
-        This code should only be called with blocks of 2880 bytes
-        Longer blocks result in non-standard checksums with carry overflow
-        Historically,  this code *was* called with larger blocks and for that
-        reason still needs to be for backward compatibility.
-        """
-        u8 = np.uint32(8)
-        u16 = np.uint32(16)
-        uFFFF = np.uint32(0xFFFF)
-
-        if data.nbytes % 2:
-            last = data[-1]
-            data = data[:-1]
-        else:
-            last = np.uint32(0)
-
-        data = data.view(">u2")
-
-        hi = sum32 >> u16
-        lo = sum32 & uFFFF
-        hi += np.add.reduce(data[0::2], dtype=np.uint64)
-        lo += np.add.reduce(data[1::2], dtype=np.uint64)
-
-        if (data.nbytes // 2) % 2:
-            lo += last << u8
-        else:
-            hi += last << u8
-
-        hicarry = hi >> u16
-        locarry = lo >> u16
-
-        while hicarry or locarry:
-            hi = (hi & uFFFF) + locarry
-            lo = (lo & uFFFF) + hicarry
-            hicarry = hi >> u16
-            locarry = lo >> u16
-
-        return (hi << u16) + lo
+        # Ensures no overflow in uint64 sum (logically, this should perhaps be
+        # 1 << 34, since we care about the number of uint32 data, not the
+        # number of bytes, but, heck, better safe than sorry.
+        blocklen = 1 << 32
+        # The cast to uint32 is not needed by the tests, and seems odd, since
+        # we explicitly use higher bits otherwise. Kept regardless since it
+        # was there before.
+        s = int(np.uint32(sum32))
+        for piece in np.split(data, range(blocklen, len(data), blocklen)):
+            if extra := piece.nbytes % 4:
+                last = bytes(piece[-extra:]) + b"\00" * (4 - extra)
+                s += int.from_bytes(last, byteorder="big", signed=False)
+                piece = piece[:-extra]
+            s += int(piece.view(">u4").sum(dtype="u8"))
+            while hi := (s >> 32):
+                s = (s & 0xFFFFFFFF) + hi
```

Comment:
If I understand correctly, what's happening here is that we're compressing information from a possibly unlimited sized `int` down to a `uint32`, but it looks to me that information is still lost if `s` initially doesn't fit in a `uint64` (specifically the loss happens on the masking operation `s & 0xFFFFFFFF`). This may never be a problem in practice, but I'm trying to make sure I'm getting the hang of it, and if I'm correct, I think this aspect should be explained in a comment too.

Reply:
> hi carries all the information of the higher bits. If those are more than 32 bits (or the sum is over 32 bits), then after the sum, s will have more than 32 bits, and hence the while runs one more time.

yes, but my comment is about `s` overflowing a `uint64` (not `uint32`). Again, very unlikely that this is ever a problem in practice, but am I right that information beyond the 64th bit would be lost ?

Reply:
Yes, it we would overflow 64 bits, that information would be lost, but given the blocklen, that cannot possibly happen, even if every `uint32=0xffffffff` (i.e., `blocklen * 0xffffffff < 0xffffffffffffffff`). See comment about blocklen.